### ISSUE TITLE
Accessibility features

### ISSUE THREAD

**Issue by kneekoo** [minetest/minetest#10666]
If anyone can help with any of these, it would be awesome.
---

Would it be possible for the game to have subtitles placed in a corner, maybe configurable? That's one of the most important accessibility features, considering hearing impaired players, or even those who cannot enable sounds for whatever reason, would still be able to play nicely if they could see that text.

As you can see, a tiny text shows you anything that you would hear, along with the description of the sound and its direction.

Basically, everything that triggers the playback of a sound also adds a subtitle in the queue. I'm not sure how long that queue can be, but my (weird) test showed 18 subtitles and it could probably go as high as the screen allows. Granted, it's extremely unlikely for so many events to happen around the player, but I did my best to trigger as many sound events (and subtitles) as possible, to test the limit.
![test for the maximum number of subtitles](https://i.imgur.com/9CRtvKZ.jpg)

A few details on the implementation can be found in [the Usage section of the Subtitles page](https://minecraft.gamepedia.com/Subtitles#Usage).

Here is the full settings page:
![Minecraft accessibility settings](https://i.imgur.com/epeSLhK.png)

Options and descriptions ([defaults listed here](https://minecraft.gamepedia.com/Options#Accessibility_Settings)):
**Narrator:** OFF, Narrates All, Narrates Chat, Narrates System
**Text Background:** Chat, Everywhere
**Chat Delay:** 0-6 seconds, with 0.1 increments
**Sneak:** Hold, Toggle
**Sprint:** Hold, Toggle

**Distortion Effects:** Strength of nausea and Nether portal screen distortion effects. At lower values, the nausea effect is replaced with a green overlay.

**FOV Effects:** Controls how much the field of view can change with speed effects.

**Accessibility Guide:** https://aka.ms/MinecraftJavaAccessibility

**Comment by SmallJoker** [comment]
Are you aware that this is repository is not about Minecraft?

**Comment by kneekoo** [comment]
@SmallJoker Are you aware that games built atop Minetest could be a lot more user-friendly if the necessary plumbing is in the engine? This is a source of inspiration for whatever would be possible to be implemented. It should make no difference that it came from Minecraft. Not everything has to be supported in Minetest, just as Minetest should not be limited to these features if anyone else has better ideas to support accessibility.

**Comment by SmallJoker** [comment]
I just wanted to ensure there's not some misunderstanding. The issue looked very much like a feature request for Minecraft, hence my uncertainty. Accessibility is surely an important thing which is still missing in Minetest.

**Comment by kneekoo** [comment]
I'm glad it's clear now, although I don't see how this could be mistaken for what you thought it was.

**Comment by LoneWolfHT** [comment]
This can already be done in games. Although a description field in sound_play and a clientside setting would make things easier

**Comment by 0-afflatus** [comment]
Can we be more precise about which events don't give enough visual feedback. I frequently play with the sound off and the only thing I'd miss is the approach of monsters. I'm all for accessibility, just subtitles seem a bit clunky.

**Comment by hecktest** [comment]
> a description field in sound_play and a clientside setting would make things easier

No, no no no no no. You can code this yourself. Soundspec is just a table and you can wrap or add any sidecar information to it if you want. Your setting can be stored on the server. Any attempt to bring this system into the engine will breed more feature requests, and any remotely competent programmer will still turn it off and do something else. A client side accessibility hint sounds good on paper, but most developers will ignore it, nobody is required to honor a client setting. It would just be a disappointment.

If you think I'm overreacting, put yourself in my shoes. So here's my noise system:
![noise](https://user-images.githubusercontent.com/42101236/100230524-1c00a880-2f26-11eb-883f-51f0333358eb.png)
It covers OP's request and a whole lot more, while being different from Minecraft and thus utterly incompatible with anything proposed here. I want to enable the messages by default and provide precise filters to tune out the "spam", and most people would have "deaf mode" off if it were implemented, so I cannot even use that kind of setting as a hint. An engine feature like this would actually end up being a negative for everyone.

**Comment by kneekoo** [comment]
> Can we be more precise about which events don't give enough visual feedback. I frequently play with the sound off and the only thing I'd miss is the approach of monsters. I'm all for accessibility, just subtitles seem a bit clunky.

While mining or moving underground, hearing stuff through walls - or being notified by subtitles - can be useful for the opportunities provided or to simply avoid stuff. The direction of the sound is equally important:
- rain fall indicates you're close to the surface, or an opening to the surface;
- bats can indicate a mine shaft, cave or ravine;
- water/lava flow can indicate potentially large spaces/openings;
- splashing or swimming reveal a nearby water body with mobs;
- footsteps immediately alert you of nearby mobs - they don't make their specific noises all the time;
- a block being broken means a sheep ate grass, likely close to the surface (you might need wool for a bed, or easy surface access);
- a creeper hissing means it's about to blow up and if you can see the direction on the sound you can quickly run in the opposite direction if you can;
- a mob falling indicates either an opening to the surface or a deep ravine;
- a mob getting hurt or dead (without falling) indicates at least two mobs or some dangerous environment;
- slime can also indicate open spaces, although they can also spawn in tight ones.

I think this is enough to prove that subtitles offer a much better sense of the surroundings, which makes for a great experience even without sounds.

@hecktest I don't know what the engine has to offer, but the most popular games based on Minetest don't have accessibility features. All I hope is that Minetest is or will be well equipped for this, and some documentation would be welcome for game developers.

**Comment by 0-afflatus** [comment]
@hecktest is basically saying this can and should be done in Lua.
@kneekoo so primary focus is to provide visual feedback for weather, water and mob activity. A system that can be generalised to translate further audio cues would be ideal.
Games, particularly 3D games aren't big on accessibility features, no. There is a reason for that, although I wouldn't call it a good one. I will consider implementing this in the games I'm designing if it can be done in Lua.

**Comment by hecktest** [comment]
This is an appeal you want to make to game developers, not engine maintainers. You might want to take this to the forums.
I already have subtitles in my game, so it definitely can be done. You don't have to rip off MC either, you can be more creative than that. The HUD is capable enough, you can display text and images anywhere you want.

**Comment by kneekoo** [comment]
@hecktest That was actually [my starting point](https://git.minetest.land/Wuzzy/MineClone2/issues/864), and the reply I got was what brought me here.

**Comment by mk-pmb** [comment]
I prefer to, as far as it's possible, to delegate the actual implementations to 3rd party accessibility programs. There's so many kinds of disabilities that it would blow project scope to re-invent a lot of wheels. Let's rather provide good APIs so others can plug in their existing solutions. (I'm currently researching how to best connect MineTest do my text-to-speech.)

**Comment by Extex101** [comment]
I suggested something like this in #9545 where it adds the ability for mods to do it

**Comment by Zughy** [comment]
Alternative approach: I've made an audio library that features accessibility options: https://gitlab.com/zughy-friends-minetest/audio_lib (CDB version doesn't feature it yet). There are a few issues that need to be addressed (check the "issues" section) but it works and you can try it on the A.E.S. server

**Comment by Zughy** [comment]
Maybe the engine should provide a general `accessibility` setting, so that mods can act accordingly to its value. For instance, right now `audio_lib` implements an accessibility option on its own, but it'd be handy to listen to the engine instead. Such setting could be a table to separate categories like `{audio, daltonism etc}`

**Comment by kneekoo** [comment]
Yes, it's best for the engine to support some basic accessibility features that the games can obey to - all optional. It makes it easier for players to set their preferences in one place, instead of each game dealing with this differently.

Once standardized, the games would choose to support whatever accessibility feature they want. How exactly this standardization happens, I guess it's up to game/mod developers to work with the Minetest team during the planning process.

**Comment by Zughy** [comment]
I'd say graphics filters e.g. for daltonism should be provided by the engine directly. On the contrary, something like visual prompts for people suffering from hear loss should just be a flag so that mods can implement their own graphical implementation (different games use different solutions)

**Comment by rubenwardy** [comment]
> Maybe the engine should provide a general `accessibility` setting, so that mods can act accordingly to its value. For instance, right now `audio_lib` implements an accessibility option on its own, but it'd be handy to listen to the engine instead. Such setting could be a table to separate categories like `{audio, daltonism etc}`

Disabilities are all different and present in different ways. So adding a setting based on _type_ of disability doesn't make sense, it's better to have settings for accessibility tools - ie: font size, subtitles, filters, etc

**Comment by Zughy** [comment]
So how would you resolve my usecase? It's visuale prompts for deaf people

**Comment by rubenwardy** [comment]
For singleplayer and client-side mods, the mods could declare a custom setting using settingtypes.txt in such a way that it appears in the "Accessibility" section as well. As in, adding `Tag: accessibility` as a line in the description

This doesn't work when the mod is server-side in multiplayer, as the client doesn't know which servers it might be joining. Accessibility does seem like a compelling usecase for client-provided client-side mods.

Another alternative would be adding support for accessibility tools to the engine

Also, I feel like adding settings for accessibility which aren't used until a mod is installed is quite opaque and confusing
